src/r_voxel.c

```diff
@@ -0,0 +1,962 @@
+	fixed_t s = finesine  [clipangle >> ANGLETOFINESHIFT];
+
+	// note the funky shifts here, to counter-balance opposite
+	// shifts in the R_PointOnSide() code.  Sigh.
+	c = c << 12;
+	s = s << 12;
+
+	node_t left;
+
+	left.x  = 0;
+	left.y  = 0;
+	left.dx = -c;
+	left.dy =  s;
+
+	if (R_PointOnSide (x1, y1, &left) &&
+	    R_PointOnSide (x2, y2, &left) &&
+	    R_PointOnSide (x3, y3, &left) &&
+	    R_PointOnSide (x4, y4, &left))
+	{
+		return false;
+	}
+
+	node_t right;
+
+	right.x  = 0;
+	right.y  = 0;
+	right.dx = -c;
+	right.dy = -s;
+
+	if (R_PointOnSide (x1, y1, &right) &&
+	    R_PointOnSide (x2, y2, &right) &&
+	    R_PointOnSide (x3, y3, &right) &&
+	    R_PointOnSide (x4, y4, &right))
+	{
+		return false;
+	}
+
+	// some part is visible
+	return true;
+}
+
+
+boolean VX_ProjectVoxel (mobj_t * thing)
+{
+	// skip the player thing we are viewing from
+	if (thing->player == viewplayer)
+		return true;
+
+	// does the voxel model exist?
+	int spr   = thing->sprite;
+	int frame = thing->frame & FF_FRAMEMASK;
+
+	struct Voxel * v = all_voxels[spr][frame];
+	if (v == NULL)
+		return false;
+
+	fixed_t gx = thing->x;
+	fixed_t gy = thing->y;
+	fixed_t gz = thing->z;
+
+	// transform the origin point
+	fixed_t tran_x = gx - viewx;
+	fixed_t tran_y = gy - viewy;
+
+	// too far away?
+	if (abs (tran_x) > VX_MAX_DIST || abs (tran_y) > VX_MAX_DIST)
+		return false;
+
+	fixed_t tx = FixedMul (tran_x, viewsin) - FixedMul (tran_y, viewcos);
+	fixed_t ty = FixedMul (tran_x, viewcos) + FixedMul (tran_y, viewsin);
+
+	fixed_t xscale;
+
+	// thing is behind view plane?  if so, ignore it
+	if (ty < VX_MINZ)
+	{
+		if (ty < -64 * FRACUNIT)
+			return true;
+
+		xscale = (15000 * FRACUNIT) - ty;
+	}
+	else
+	{
+		// too far off the side?  if so, ignore it
+		if (ty > (64 * FRACUNIT) && abs(tx) / 4 > ty)
+			return true;
+
+		xscale = FixedDiv (projection, ty);
+	}
+
+	angle_t angle = thing->angle;
+
+	switch (VX_RotateModeForThing (thing))
+	{
+		case 1:
+			angle = viewangle + ANG180;
+			break;
+
+		case 2:
+			angle = R_PointToAngle (gx, gy) + ANG180;
+			break;
+	}
+
+	angle_t ang2 = ANG180 - viewangle + angle;
+
+	// unit vector from top-left corner toward top-right corner
+	fixed_t c = finecosine[ang2 >> ANGLETOFINESHIFT];
+	fixed_t s = finesine  [ang2 >> ANGLETOFINESHIFT];
+
+	// compute coord of top-left corner (transformed)
+	fixed_t xp = v->x_pivot;
+	fixed_t yp = v->y_pivot;
+
+	fixed_t TL_x = tx - FixedMul (xp, c) - FixedMul (yp, s);
+	fixed_t TL_y = ty - FixedMul (xp, s) + FixedMul (yp, c);
+
+	// compute X coords of the other corners
+	fixed_t xs = v->x_size;
+	fixed_t ys = v->y_size;
+
+	fixed_t BL_x = TL_x + (ys * s);
+	fixed_t BL_y = TL_y - (ys * c);
+
+	fixed_t TR_x = TL_x + (xs * c);
+	fixed_t TR_y = TL_y + (xs * s);
+
+	fixed_t BR_x = BL_x + (xs * c);
+	fixed_t BR_y = BL_y + (xs * s);
+
+	// is it outside the view frustum?
+	if (! VX_CheckFrustum (TL_x, TL_y, BL_x, BL_y, BR_x, BR_y, TR_x, TR_y))
```

Comment:
I fixed crashes, everything works except the widescreen clipping. The `VX_CheckFrustrum()` function works only for 4:3 aspect ratio.

Reply:
This is one way to fix widescreen:

<details><summary>clip angle</summary>

```diff
diff --git a/src/r_main.c b/src/r_main.c
index b56e486..16dccb1 100644
--- a/src/r_main.c
+++ b/src/r_main.c
@@ -489,6 +489,7 @@ void R_ExecuteSetViewSize (void)
   R_InitBuffer(scaledviewwidth, scaledviewheight);       // killough 11/98
         
   R_InitTextureMapping();
+  VX_UpdateClipAngle();
     
   // psprite scales
   pspritescale = FixedDiv(viewwidth_nonwide, ORIGWIDTH);       // killough 11/98
```

```diff
diff --git a/src/r_voxel.c b/src/r_voxel.c
index bae6d05..74ce9dd 100644
--- a/src/r_voxel.c
+++ b/src/r_voxel.c
@@ -382,6 +382,15 @@ static int VX_RotateModeForThing (mobj_t * thing)
 }
 
 
+#define VX_PI 3.14159265358979323846
+static angle_t vx_clipangle;
+
+void VX_UpdateClipAngle(void)
+{
+	const double fovrad = 2 * atan(SCREENWIDTH / (1.2 * SCREENHEIGHT) * 3 / 4);
+	vx_clipangle = fovrad / VX_PI * ANG180 - ANG90;
+}
+
 static boolean VX_CheckFrustum (fixed_t x1, fixed_t y1, fixed_t x2, fixed_t y2,
                          fixed_t x3, fixed_t y3, fixed_t x4, fixed_t y4)
 {
@@ -389,8 +398,8 @@ static boolean VX_CheckFrustum (fixed_t x1, fixed_t y1, fixed_t x2, fixed_t y2,
 	if (y1 < VX_MINZ && y2 < VX_MINZ && y3 < VX_MINZ && y4 < VX_MINZ)
 		return false;
 
-	fixed_t c = finecosine[clipangle >> ANGLETOFINESHIFT];
-	fixed_t s = finesine  [clipangle >> ANGLETOFINESHIFT];
+	fixed_t c = finecosine[vx_clipangle >> ANGLETOFINESHIFT];
+	fixed_t s = finesine  [vx_clipangle >> ANGLETOFINESHIFT];
 
 	// note the funky shifts here, to counter-balance opposite
 	// shifts in the R_PointOnSide() code.  Sigh.
```

```diff
diff --git a/src/r_voxel.h b/src/r_voxel.h
index 8ec92d1..e6828dc 100644
--- a/src/r_voxel.h
+++ b/src/r_voxel.h
@@ -27,4 +27,6 @@ boolean VX_ProjectVoxel (mobj_t * thing);
 
 void VX_DrawVoxel (vissprite_t * vis);
 
+void VX_UpdateClipAngle(void);
+
 #endif  /* __R_VOXEL__ */
```

</details>

Reply:
Thanks for idea, this seems to work:

<details>

```diff
diff --git a/src/i_video.c b/src/i_video.c
index 89263d91..058ef4f8 100644
--- a/src/i_video.c
+++ b/src/i_video.c
@@ -45,6 +45,7 @@
 int SCREENWIDTH, SCREENHEIGHT;
 int NONWIDEWIDTH; // [crispy] non-widescreen SCREENWIDTH
 int WIDESCREENDELTA; // [crispy] horizontal widescreen offset
+int FOV;
 
 boolean use_vsync;  // killough 2/8/98: controls whether vsync is called
 int hires, default_hires;      // killough 11/98
@@ -958,6 +959,13 @@ void I_GetScreenDimensions(void)
     }
 
     WIDESCREENDELTA = (SCREENWIDTH - NONWIDEWIDTH) / 2;
+
+    // The general equation is as follows:
+    // y = mx + b -> fov = (75/2) * ratio + 40
+    // This gives 90 for 4:3, 100 for 16:10, and 106 for 16:9.
+
+    double ratio = (double)SCREENWIDTH / (double)ah;
+    FOV = (int)((75.0/2.0) * ratio + 40.0);
 }
 
 static void CreateUpscaledTexture(boolean force)
diff --git a/src/i_video.h b/src/i_video.h
index 6afa4399..dbf67de1 100644
--- a/src/i_video.h
+++ b/src/i_video.h
@@ -28,6 +28,7 @@ extern int SCREENWIDTH;
 extern int SCREENHEIGHT;
 extern int NONWIDEWIDTH; // [crispy] non-widescreen SCREENWIDTH
 extern int WIDESCREENDELTA; // [crispy] horizontal widescreen offset
+extern int FOV;
 void I_GetScreenDimensions (void); // [crispy] re-calculate WIDESCREENDELTA
 
 enum
diff --git a/src/r_main.c b/src/r_main.c
index b56e486d..d20789d1 100644
--- a/src/r_main.c
+++ b/src/r_main.c
@@ -57,6 +57,7 @@ fixed_t  viewheightfrac; // [FG] sprite clipping optimizations
 //
 
 angle_t clipangle;
+angle_t vx_clipangle;
 
 // The viewangletox[viewangle + FINEANGLES/4] lookup
 // maps the visible view angles to screen X coordinates,
@@ -305,6 +306,8 @@ static void R_InitTextureMapping (void)
         viewangletox[i] = viewwidth;
         
   clipangle = xtoviewangle[0];
+
+  vx_clipangle = clipangle - (FOV - 90) * ANG1;
 }
 
 //
diff --git a/src/r_state.h b/src/r_state.h
index e6668610..69ca7abe 100644
--- a/src/r_state.h
+++ b/src/r_state.h
@@ -96,6 +96,7 @@ extern fixed_t          viewz;
 extern angle_t          viewangle;
 extern player_t         *viewplayer;
 extern angle_t          clipangle;
+extern angle_t          vx_clipangle;
 extern int              viewangletox[FINEANGLES/2];
 extern angle_t          xtoviewangle[MAX_SCREENWIDTH+1];  // killough 2/8/98
 extern fixed_t          rw_distance;
diff --git a/src/r_voxel.c b/src/r_voxel.c
index bae6d052..ff05dee5 100644
--- a/src/r_voxel.c
+++ b/src/r_voxel.c
@@ -389,8 +389,8 @@ static boolean VX_CheckFrustum (fixed_t x1, fixed_t y1, fixed_t x2, fixed_t y2,
 	if (y1 < VX_MINZ && y2 < VX_MINZ && y3 < VX_MINZ && y4 < VX_MINZ)
 		return false;
 
-	fixed_t c = finecosine[clipangle >> ANGLETOFINESHIFT];
-	fixed_t s = finesine  [clipangle >> ANGLETOFINESHIFT];
+	fixed_t c = finecosine[vx_clipangle >> ANGLETOFINESHIFT];
+	fixed_t s = finesine  [vx_clipangle >> ANGLETOFINESHIFT];
 
 	// note the funky shifts here, to counter-balance opposite
 	// shifts in the R_PointOnSide() code.  Sigh.
```
</details>

I think we need to add a custom FOV anyway.

Reply:
Unfortunately it seems to flicker: [allmonsters.zip](https://github.com/fabiangreffrath/woof/files/13316051/allmonsters.zip)
```
woof -file allmonsters.wad -playdemo allmonsters
```

Reply:
> Unfortunately it seems to flicker: [allmonsters.zip](https://github.com/fabiangreffrath/woof/files/13316051/allmonsters.zip)

It flickers also with normal `clipangle` when widescreen mode is off. I think there are problems with very large models.

Reply:
Okay, it looks good to me other than that.

Reply:
Not sure if he has an account on github, but maybe we should ask Andrew Apted for his input?

Reply:
Flickering disappears in 200p mode. I missed `hires` somewhere.

Reply:
Fixed `hires` flickering: https://github.com/fabiangreffrath/woof/pull/1249/commits/89e71c83a6531310109cd73dc08ba4ddcd9714e3 [allmonsters.zip](https://github.com/fabiangreffrath/woof/files/13316051/allmonsters.zip) demo works

Reply:
Awesome, thank you! How about the "view size" issue?

Reply:
> How about the "view size" issue?

Don't know yet. I guess we need to use `viewwindowx` and `viewwindowy` offsets.

Reply:
> Fixed `hires` flickering: [89e71c8](https://github.com/fabiangreffrath/woof/commit/89e71c83a6531310109cd73dc08ba4ddcd9714e3) [allmonsters.zip](https://github.com/fabiangreffrath/woof/files/13316051/allmonsters.zip) demo works

I did a quick test and it looks good.

Reply:
> How about the "view size" issue?

Fixed in https://github.com/fabiangreffrath/woof/pull/1249/commits/258742b646f3a6b88cb02e67b6c4008cadcaf062

Reply:
Awesome!